core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -576,96 +583,252 @@ public void addPhysicallyWrittenRows(long rows) {
+            LOG.info().$("partition is already attached [path=").$(path).I$();
+            // TODO: potentially we can merge with existing data
+            return StatusCode.PARTITION_ALREADY_ATTACHED;
         }
 
         boolean rollbackRename = false;
         try {
-            setPathForPartition(path, partitionBy, timestamp, false);
-            if (!ff.exists(path.$())) {
-                setPathForPartition(other, partitionBy, timestamp, false);
-                other.put(DETACHED_DIR_MARKER);
-
-                if (ff.exists(other.$())) {
-                    if (ff.rename(other, path) == Files.FILES_RENAME_OK) {
-                        rollbackRename = true;
-                        LOG.info().$("moved partition dir: ").$(other).$(" to ").$(path).$();
-                    } else {
-                        throw CairoException.instance(ff.errno()).put("File system error on trying to rename [from=")
-                                .put(other).put(",to=").put(path).put(']');
-                    }
-                }
+            if (detachedPath == null) {
+                other2 = new Path();
+                detachedPath = new Path().of(configuration.getDetachedRoot()).concat(tableName);
+                detachedRootLen = detachedPath.length();
             }
 
-            if (ff.exists(path.$())) {
-                // find out lo, hi ranges of partition attached as well as size
-                final long partitionSize = readPartitionSizeMinMax(ff, path, timestampCol, tempMem16b, timestamp);
-                if (partitionSize > 0) {
+            // final name of partition folder after attach
+            setPathForPartition(path.trimTo(rootLen), partitionBy, timestamp, false);
+            path.slash$();
+
+            long partitionSize = 0;
+            if (!ff.exists(path)) {
+                setPathForPartition(detachedPath.trimTo(detachedRootLen), partitionBy, timestamp, false);
+                detachedPath.put(ATTACHABLE_DIR_MARKER).slash$();
+                if (ff.exists(detachedPath)) {
                     if (inTransaction()) {
                         LOG.info().$("committing open transaction before applying attach partition command [table=").$(tableName)
-                                .$(",partition=").$ts(timestamp).I$();
+                                .$(", partition=").$ts(timestamp).I$();
                         commit();
                     }
+
+                    CharSequence timestampColName = metadata.getColumnQuick(metadata.getTimestampIndex()).getName();
+                    partitionSize = readPartitionSizeMinMax(ff, detachedPath, timestampColName, tempMem16b, timestamp);
+                    if (partitionSize == -1) {
+                        throw CairoException.detachedMetadataMismatch("missing_column=" + timestampColName);
+                    }
+
+                    // detached metadata files validation
                     if (validateDataFiles) {
```

Comment:
```
if (paritionSize > 0 && validateDataFiles) {
```

Reply:
done in [47a14c2](https://github.com/questdb/questdb/pull/2280/commits/47a14c2e53d813b5612f72cf047cde353adfd205)